libs/metall/src/dice/metall/detail/file.hpp

```diff
@@ -281,54 +279,165 @@ inline bool copy_file_dense(const std::filesystem::path &source_path,
 }
 
 #ifdef __linux__
-inline bool prepare_file_copy_linux(const std::filesystem::path &source_path,
-                                    const std::filesystem::path &destination_path,
-                                    int *src,
-                                    int *dst) {
+/**
+ * @brief Prepares a file copy from source_path to destination_path
+ *      by opening/creating the relevant files and setting appropriate permissions.
+ *
+ * @param source_path path to source file
+ * @param destination_path desired path of destination file
+ * @param src out parameter for the file descriptor opened for source_path (will be opened read only)
+ * @param dst out parameter for the file descriptor opened for destination_path (will be opened write only)
+ * @return on success: size of source file as obtained by ::fstat. on failure: -1
+ *
+ * @warning if the function fails the user must not use the obtained src and dst file descriptors in any way
+ *      (they don't need to be closed)
+ */
+inline off_t prepare_file_copy_linux(const std::filesystem::path &source_path,
+                                     const std::filesystem::path &destination_path,
+                                     int *src,
+                                     int *dst) {
   *src = ::open(source_path.c_str(), O_RDONLY);
   if (*src == -1) {
     METALL_ERRNO_ERROR("Unable to open {}", source_path.c_str());
-    return false;
+    return -1;
   }
 
   struct stat st;
   if (::fstat(*src, &st) == -1) {
     METALL_ERRNO_ERROR("Unable to stat {}", source_path.c_str());
     os_close(*src);
-    return false;
+    return -1;
   }
 
   *dst = ::open(destination_path.c_str(), O_WRONLY | O_CREAT | O_TRUNC, st.st_mode);
   if (*dst == -1) {
     METALL_ERRNO_ERROR("Unable to open {}", destination_path.c_str());
     os_close(*src);
+    return -1;
+  }
+
+  return st.st_size;
+}
+
+/**
+ * Creates a hole of size size after the current cursor of fd.
+ * Moves cursor of fd behind the created hole.
+ *
+ * @param fd file descriptor
+ * @param size size of to-be-created hole
+ * @return if creation was successful
+ *
+ * Relevant man pages:
+ *      https://www.man7.org/linux/man-pages/man2/lseek.2.html
+ *      https://man7.org/linux/man-pages/man2/fallocate.2.html
+ */
+inline bool create_hole_linux(int fd, off_t size) {
+  if (size == 0) {
+    return true;
+  }
+
+  // Seek size bytes past the current position
+  off_t new_file_end = ::lseek(fd, size, SEEK_CUR);
```

Comment:
Doing it this way because `cp` also does this. There may be a reason for that, or not i don't know tbh.

Modifying the file does not matter in the case of failure because we start from scratch anyways when this fails in any way.

Reply:
I would simply add a comment to the docstring that the position in the file is guaranteed in case of failure.